front: fix NGE import error when a TrainrunSection is reversed

emersion opened the pull request:
NGE doesn't guarantee that trainrun sections will always follow the

    [source → target] [source → target] …

order. The following are valid NGE trainruns:

    [source → target] [target → source]
    [target → source] [target → source]
    [target → source] [source → target]

Introduce a more robust graph search algorithm which accommodates for these cases and is closer to NGE's TrainrunIterator, by tracking the current node in addition to the current section.

To test:

- Open https://nge.flatland.cloud/
- Create a train run from LTH to ZF
- Extend the train run by clicking from ZG to ZF (*not* from ZF to ZG)
- Export as JSON (result:  [networkGraphic.json](https://github.com/user-attachments/files/20779478networkGraphic.1.json))
- Import into OSRD (failed before, should work now)

emersion committed:
front: fix NGE import error when a TrainrunSection is reversed

NGE doesn't guarantee that trainrun sections will always follow the

    [source → target] [source → target] …

order. The following are valid NGE trainruns:

    [source → target] [target → source]
    [target → source] [target → source]
    [target → source] [source → target]

Introduce a more robust graph search algorithm which accommodates
for these cases and is closer to NGE's TrainrunIterator, by tracking
the current node in addition to the current section.

Also prepare for future work to handle non-continuous trainruns.

Signed-off-by: Simon Ser <contact@emersion.fr>

louisgreiner commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -74,58 +82,107 @@ const getTrainrunSectionsByTrainrunId = (netzgrafikDto: NetzgrafikDto, trainrunI
   }
 
-  return orderedSections;
+  return orderedSectionPaths[0];
```
Wonderful, non-continuous trainruns will be fixed easier now

louisgreiner reviewed (approved):
Tested, it works well for import, and has the same behavior when playing with NGE as before.

We can also think on noticing the user when one tries to create non-continuous trainrun, I created [this issue](https://github.com/osrd-project/osrd-confidential/issues/1132)

emersion commented:
I've split off the hacks in a separate temporary PR: https://github.com/OpenRailAssociation/osrd/pull/12212

Also debugging for duplicate nodes: https://github.com/OpenRailAssociation/osrd/compare/emr/nge-import-log

shenriotpro commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -74,58 +82,107 @@ const getTrainrunSectionsByTrainrunId = (netzgrafikDto: NetzgrafikDto, trainrunI
   }
 
-  return orderedSections;
+  // TODO: some train runs are made up of multiple parts. We should create one
```
IMO we should raise an error, but I'm not a stakeholder :)

Math-R replied:
Yes agree with that

emersion replied:
Right. I originally kept the error, and then removed it because the TODO will soon get resolved (in an upcoming patch) and that at least allows us to import all trainruns from RNE's large file. But with the hack commits linked above, we can workaround this easily for now - so let's re-introduce the error for now.

shenriotpro commented:
What's the context? Are we trying to import schedules that don't come from OSRD?

shenriotpro commented:
> To test:
> 
> * Open https://nge.flatland.cloud/
> * Create a train run from LTH to ZF
> * Extend the train run by clicking from ZG to ZF (_not_ from ZF to ZG)
> * Export as JSON (result:  [networkGraphic.json](https://github.com/user-attachments/files/20779478networkGraphic.1.json))
> * Import into OSRD (failed before, should work now)

Perhaps this could be turned into a unit test?

louisgreiner commented:
@shenriotpro exactly, the context is that when we want to import NGE files not coming from OSRD. The issue is that NGE allows to draw trainrun sections weirdly order. This PR intend to ensure logical order

Math-R commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -74,58 +82,107 @@ const getTrainrunSectionsByTrainrunId = (netzgrafikDto: NetzgrafikDto, trainrunI
-    if (transition) {
-      const prevPortId =
-        transition.port1Id === section.sourcePortId ? transition.port2Id : transition.port1Id;
-      sectionsByPrevTargetPortId.set(prevPortId, section);
+    const targetNode = getNodeById(netzgrafikDto.nodes, section.targetNodeId)!;
+
+    const sourceConnectedPortId = findConnectedPortId(sourceNode, section.sourcePortId);
+    const targetConnectedPortId = findConnectedPortId(targetNode, section.targetPortId);
+
+    if (sourceConnectedPortId !== null) {
+      sectionsByConnectedTargetPortId.set(sourceConnectedPortId, section);
     } else {
-      departureSection = section;
+      leafSectionsAndNodes.push([section, section.sourceNodeId]);
+    }
+    if (targetConnectedPortId !== null) {
+      sectionsByConnectedSourcePortId.set(targetConnectedPortId, section);
+    } else {
+      leafSectionsAndNodes.push([section, section.targetNodeId]);
     }
```
I noticed that if both `sourceConnectedPortId `and `targetConnectedPortId `are null,  this section gets added twice (once for source, once for target).
Can this lead to unexpected behaviour in the case where there are not connected ?

emersion replied:
This can happen when a train run has a single section. In that case, the lone section has no neighbor so both of these variables will be null.

This won't lead to unexpected behavior, because in the loop below we check whether we've already seen a section before iterating over a path. So we'll start walking on the section from one end, and then skip the other one.

emersion committed:
front: add out-of-order source/target tests for convertNgeDtoToOsrd()

Ensure the import logic works fine with out-of-order source/target
train run sections.

NGE and OSRD dump files are stored in separate JSON files so that
they can easily be imported/exported into OSRD/NGE for inspection
or patching.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion commented:
> Perhaps this could be turned into a unit test?

Done!

emersion commented:
I've noticed that the graph traversal was too strict and not grouping all sections correctly with two maps. Simplifying to a single `sectionsByConnectedPortId` map fixes this. It makes sense because transitions don't care about source/target ordering, and a port is connected to at most one section via a transition.

shenriotpro commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -74,58 +82,109 @@ const getTrainrunSectionsByTrainrunId = (netzgrafikDto: NetzgrafikDto, trainrunI
+        nodeId = section.sourceNodeId;
+        section = sectionsByConnectedPortId.get(section.sourcePortId);
+      } else {
+        throw new Error('Section is disconnected from previous section');
```
nit: is that reachable?

emersion replied:
Yes, this error acts as an unreachable guard. There is no good way to mark a branch as unreachable in JavaScript AFAIK. I find it important to fail hard if assumptions like these aren't met. There is `console.assert()` but it's banned by ESLint rules and doesn't stop execution.

shenriotpro replied:
Thanks, can we add a comment about that? The error message is not very clear to me

emersion replied:
Added the following:

```ts
        // Unreachable: the previous section's end node should be either this
        // node's source or target
```

shenriotpro commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -74,58 +82,109 @@ const getTrainrunSectionsByTrainrunId = (netzgrafikDto: NetzgrafikDto, trainrunI
+    let section: TrainrunSectionDto | undefined = startSection;
+    let nodeId = startNodeId;
+    const orderedSections = [];
+    while (section) {
```
nit: now that we raise for disconnects, we can slightly simplify the code, by just traversing once, storing one path, and checking the length

emersion replied:
We really want to handle multiple paths per train run differently in the near future, so I'd rather not remove some code that we'll need to re-introduce very soon.

shenriotpro replied:
Alright

Math-R reviewed (approved):
lgtm !
